### ISSUE TITLE
Add support for Content-Security-Policy

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/view#31]
We should add advanced support for `Content-Security-Policy` directives:

- [ ] Sign inline scripts using a hash.
- [ ] Only allow scripts that are included via `registerJs`, `registerJsFile`.
- [ ] Research what else we should support (CSP has options for framing as well as font loading).
- [ ] Support CSP for asset bundles

**Comment by samdark** [comment]
Any good reading on the subject?

**Comment by terabytesoftw** [comment]
I think it could be done in two ways via **.htaccess** and via **Html :: tag ('meta')**

**Comment by SamMousa** [comment]
https://websec.be/blog/cspstrictdynamic/

@terabytesoftw as a PHP framework `.htaccess` is not part of our stack.
CSP can be set via `Meta` tag and via headers.

**Comment by samdark** [comment]
@SamMousa I've checked it. Seems what's proposed with hashes still require extensive fallback usage. While hashes could be automated, I'm not sure about fallback.

If it could be done in a universal way, I am all for implementing it as part of assets concept.

**Comment by terabytesoftw** [comment]
in `Yiisoft/Web` in `SapiEmiter.php` in emit you can add:

```
$response = $response
    ->withHeader('Access-Control-Allow-Credentials', 'true')
```
and the remaining headers.

**Comment by samdark** [comment]
No, that is not the place to do it. SapiEmitter is just "echoing" what is contained in the response object as is. If you need extra headers, these should be added in a middleware.

**Comment by SamMousa** [comment]
Since this is the view component, would it make sense to always use a meta tag? 
(That way we can keep the whole implementation local to view rendering)

Regarding script src verification, we could use  a random nonce for every request; this would be really simple to implement.

**Comment by SamMousa** [comment]
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src

**Comment by samdark** [comment]
> Since this is the view component, would it make sense to always use a meta tag?
> (That way we can keep the whole implementation local to view rendering)

OK.

> Regarding script src verification, we could use a random nonce for every request; this would be really simple to implement.

Fine.
